File: selenium-python/pages/base_page.py
"""Base class for all page objects.

Two shared, per-driver handlers are resolved once in ``__init__`` and inherited
by every page object — subclasses never construct a ``WebDriverWait``.
:class:`~pages.wait_handlers.WaitHandlers` decides how long to wait
(``short_wait`` / ``wait`` / ``long_wait``);
:class:`~pages.element_handler.ElementHandler` decides what to do with the
element. Page methods express intent (``is_visible``, ``click``,
``await_true``) and pick a tier.
"""

from __future__ import annotations

from typing import Any, Callable, List, Optional

from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.remote.webelement import WebElement
from selenium.webdriver.support.ui import WebDriverWait

from pages.element_handler import ElementHandler
from pages.wait_handlers import Locator, WaitHandlers


class BasePage:
    """Base class for all page objects."""

    def __init__(self, driver: WebDriver, base_url: str = "") -> None:
        self.driver = driver
        self.base_url = base_url
        # Shared, per-driver handlers — also used directly by step definitions.
        self.waits = WaitHandlers.for_driver(driver)      # how long to wait
        self.elements = ElementHandler.for_driver(driver)  # what to do with the element
        self.short_wait = self.waits.short_wait      # 5 s  — negative checks, animation
        self.wait = self.waits.default_wait          # 10 s — element interaction
        self.long_wait = self.waits.long_wait        # 15 s — page load, SPA hydration

    def new_wait(self, timeout: float) -> WebDriverWait:
        """Handler for a timeout the three tiers don't cover."""
        return self.waits.custom(timeout)

    def open(self, path: str = "") -> None:
        self.driver.get(f"{self.base_url}{path}")

    # ── Element access ────────────────────────────────────────────────────

    def find(self, locator: Locator, handler: Optional[WebDriverWait] = None) -> WebElement:
        return self.elements.present(locator, handler)

    def find_all(self, locator: Locator) -> List[WebElement]:
        return self.elements.all(locator)

    def wait_visible(self, locator: Locator, handler: Optional[WebDriverWait] = None) -> WebElement:
        return self.elements.visible(locator, handler)

    def click(self, locator: Locator, handler: Optional[WebDriverWait] = None) -> None:
        self.elements.click(locator, handler)

    def type_text(self, locator: Locator, text: str) -> None:
        self.elements.type_text(locator, text)

    def get_text(self, locator: Locator) -> str:
        return self.elements.text(locator)

    # ── Boolean-returning waits ───────────────────────────────────────────

    def is_visible(self, locator: Locator, handler: Optional[WebDriverWait] = None) -> bool:
        return self.elements.is_visible(locator, handler)

    def is_clickable(self, locator: Locator, handler: Optional[WebDriverWait] = None) -> bool:
        return self.elements.is_clickable(locator, handler)

    def any_visible(
        self, locator: Locator, handler: Optional[WebDriverWait] = None
    ) -> WebElement:
        """First displayed match — use when the locator can match a hidden duplicate."""
        return self.elements.any_visible(locator, handler)

    def is_any_visible(
        self, locator: Locator, handler: Optional[WebDriverWait] = None
    ) -> bool:
        return self.elements.is_any_visible(locator, handler)

    def click_any_visible(
        self, locator: Locator, handler: Optional[WebDriverWait] = None
    ) -> None:
        self.elements.click_any_visible(locator, handler)

    def await_true(
        self,
        condition: Callable[[WebDriver], Any],
        handler: Optional[WebDriverWait] = None,
    ) -> bool:
        return self.waits.await_true(condition, handler)

    def await_not_displayed(
        self, locator: Locator, handler: Optional[WebDriverWait] = None
    ) -> bool:
        return self.elements.await_not_displayed(locator, handler)

    # ── Scripting ─────────────────────────────────────────────────────────

    def execute_script(self, script: str, *args: Any) -> Any:
        return self.elements.execute_script(script, *args)

    def wait_for_document_ready(self, handler: Optional[WebDriverWait] = None) -> None:
        self.elements.document_ready(handler)

    @property
    def title(self) -> str:
        return self.driver.title

    @property
    def current_url(self) -> str:
        return self.driver.current_url

File: selenium-java/src/test/java/com/example/tests/pages/BasePage.java
package com.example.tests.pages;

import com.example.tests.support.ElementHandler;
import com.example.tests.support.WaitHandlers;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.List;
import java.util.function.Function;

/**
 * Base class for all page objects.
 *
 * <p>Two shared, per-driver handlers are resolved once in this constructor and
 * inherited by every page object — subclasses never construct a
 * {@link WebDriverWait}. {@link WaitHandlers} decides how long to wait
 * ({@link #shortWait} / {@link #wait} / {@link #longWait});
 * {@link ElementHandler} decides what to do with the element. Page methods
 * express intent ({@code isVisible}, {@code click}, {@code awaitTrue}) and
 * pick a tier.
 */
public abstract class BasePage {

    protected final WebDriver driver;
    protected final String baseUrl;

    /** Wait tiers — shared with step definitions via {@code WaitHandlers.forDriver}. */
    protected final WaitHandlers waits;
    /** Element interaction on top of those tiers. */
    protected final ElementHandler elements;

    /** 5 s — negative checks and dropdown/animation polling. */
    protected final WebDriverWait shortWait;
    /** 10 s — the default for element interaction and visibility. */
    protected final WebDriverWait wait;
    /** 15 s — page load and SPA hydration. */
    protected final WebDriverWait longWait;

    protected BasePage(WebDriver driver, String baseUrl) {
        this.driver = driver;
        this.baseUrl = baseUrl == null ? "" : baseUrl;
        this.waits = WaitHandlers.forDriver(driver);
        this.elements = ElementHandler.forDriver(driver);
        this.shortWait = waits.shortWait();
        this.wait = waits.defaultWait();
        this.longWait = waits.longWait();
    }

    /** Handler for a timeout the three tiers don't cover. */
    protected WebDriverWait newWait(Duration timeout) {
        return waits.custom(timeout);
    }

    public void open(String path) {
        driver.get(baseUrl + (path == null ? "" : path));
    }

    // ── Element access ────────────────────────────────────────────────────

    protected WebElement find(By locator) {
        return elements.present(locator);
    }

    protected WebElement find(By locator, WebDriverWait handler) {
        return elements.present(locator, handler);
    }

    protected WebElement waitVisible(By locator) {
        return elements.visible(locator);
    }

    protected WebElement waitVisible(By locator, WebDriverWait handler) {
        return elements.visible(locator, handler);
    }

    protected void click(By locator) {
        elements.click(locator);
    }

    protected void click(By locator, WebDriverWait handler) {
        elements.click(locator, handler);
    }

    protected void type(By locator, String text) {
        elements.type(locator, text);
    }

    protected String text(By locator) {
        return elements.text(locator);
    }

    protected List<WebElement> findAll(By locator) {
        return elements.all(locator);
    }

    // ── Boolean-returning waits ───────────────────────────────────────────

    protected boolean isVisible(By locator) {
        return elements.isVisible(locator);
    }

    protected boolean isVisible(By locator, WebDriverWait handler) {
        return elements.isVisible(locator, handler);
    }

    protected boolean isClickable(By locator, WebDriverWait handler) {
        return elements.isClickable(locator, handler);
    }

    /** First displayed match — use when the locator can also match a hidden duplicate. */
    protected WebElement anyVisible(By locator, WebDriverWait handler) {
        return elements.anyVisible(locator, handler);
    }

    protected boolean isAnyVisible(By locator) {
        return elements.isAnyVisible(locator);
    }

    protected boolean isAnyVisible(By locator, WebDriverWait handler) {
        return elements.isAnyVisible(locator, handler);
    }

    protected void clickAnyVisible(By locator) {
        elements.clickAnyVisible(locator);
    }

    protected void clickAnyVisible(By locator, WebDriverWait handler) {
        elements.clickAnyVisible(locator, handler);
    }

    protected boolean awaitTrue(Function<WebDriver, Boolean> condition, WebDriverWait handler) {
        return waits.awaitTrue(condition, handler);
    }

    protected boolean awaitNotDisplayed(By locator, WebDriverWait handler) {
        return elements.awaitNotDisplayed(locator, handler);
    }

    // ── Scripting ─────────────────────────────────────────────────────────

    protected Object executeScript(String script, Object... args) {
        return elements.executeScript(script, args);
    }

    /** Waits for {@code document.readyState == "complete"}. */
    protected void waitForDocumentReady(WebDriverWait handler) {
        elements.documentReady(handler);
    }

    public String title() {
        return driver.getTitle();
    }

    public String currentUrl() {
        return driver.getCurrentUrl();
    }
}
